lib/buffer.js

```diff
@@ -557,21 +557,39 @@ Buffer.prototype.compare = function compare(target,
 // - encoding - an optional encoding, relevant is val is a string
 // - dir - true for indexOf, false for lastIndexOf
 function bidirectionalIndexOf(buffer, val, byteOffset, encoding, dir) {
-  if (typeof byteOffset === 'string') {
-    encoding = byteOffset;
-    byteOffset = undefined;
-  } else if (byteOffset > 0x7fffffff) {
-    byteOffset = 0x7fffffff;
-  } else if (byteOffset < -0x80000000) {
-    byteOffset = -0x80000000;
-  }
-  byteOffset = +byteOffset;  // Coerce to Number.
-  if (isNaN(byteOffset)) {
-    // If the offset is undefined, null, NaN, "foo", etc, search whole buffer.
-    byteOffset = dir ? 0 : (buffer.length - 1);
-  }
+
   dir = !!dir;  // Cast to bool.
 
+  if (typeof byteOffset === 'number') {
+    // Make the general case fast: byteOffset is a number
+    if (byteOffset >= -0x80000000 && byteOffset <= 0x7fffffff) {
+	//general case
+    } else if (byteOffset > 0x7fffffff) {
+      byteOffset = 0x7fffffff;
+    } else if (byteOffset < -0x80000000) {
+      byteOffset = -0x80000000;
+    } else {
+      // No domain value possible remained except NaN
+      byteOffset = dir ? 0 : (buffer.length - 1);
+    }
+  } else {
+    if (typeof byteOffset === 'string') {
+      //if byteOffset is a string, bidirectionalIndexOf was called
+      //with fewer parameters
+      //we'll try to guess the correct parameters
+      encoding = byteOffset;
+      dir = true;
+      byteOffset = 0;
+    } else {
+      // special cases, the byteOffset is an object
+      // this might be null, []. {} or other variations
+      byteOffset = +byteOffset;
+      if (isNaN(byteOffset)) {
```

Comment:
Can't we just replace `isNaN()` here with `byteOffset !== byteOffset` ? It should be safe since `byteOffset` will never be `undefined` at this point.